src/Adapter/AbstractObjectModelRepository.php

```diff
@@ -74,28 +80,82 @@ protected function assertObjectModelExists(int $id, string $objectTableName, str
      */
     protected function getObjectModel(int $id, string $objectModelClass, string $exceptionClass): ObjectModel
     {
-        try {
-            $objectModel = new $objectModelClass($id);
+        return $this->fetchObjectModel($id, $objectModelClass, $exceptionClass, null);
+    }
 
-            if ((int) $objectModel->id !== $id) {
-                throw new $exceptionClass(sprintf('%s #%d was not found', $objectModelClass, $id));
-            }
-        } catch (PrestaShopException $e) {
-            throw new CoreException(
-                sprintf(
-                    'Error occurred when trying to get %s #%d [%s]',
-                    $objectModelClass,
-                    $id,
-                    $e->getMessage()
-                ),
-                0,
-                $e
-            );
-        }
+    /**
+     * @param int $id
+     * @param string $objectModelClass
+     * @param string $exceptionClass
+     *
+     * @return ObjectModel
+     *
+     * @throws CoreException
+     */
+    protected function getObjectModelForShop(int $id, string $objectModelClass, string $exceptionClass, ShopId $shopId): ObjectModel
```

Comment:
Split the multishop functions and move them into a dedicated multishop abstract
The ProductRepository will have to extend this new abstract, temporarily we will have two product repositories one multishop one "classic" the aim is to replace the classic one with the multishop one piece by piece and finally only keep the multishop one (which will be renamed into ProductRepository at that final point)